### PR TITLE
Improve handling HTML special cases

### PR DESCRIPTION
**Pull request by jelmervdl**
Draft for now to just capture ongoing improvements

Change log
- If markup appears in any of the tokens that make up a word, spread that markup to the rest of the word. Even if alignment scores for that token are lower than other tokens. Fixes the link "disappearing" from "Ceasar" in `<p><b>Cayo</b> o <b>Gayo Trebacio Testa</b> (m. ca. 4) nacido probablemente en <a href="/wiki/Velia" title="Velia">Velia</a>, <a href="/wiki/Basilicata" title="Basilicata">Lucania</a>, fue uno de los abogados más prominentes de su época. Amigo de <a href="/wiki/Cicer%C3%B3n" title="Cicerón">Cicerón</a>, fue recomendado por este a <a href="/wiki/Julio_C%C3%A9sar" title="Julio César">César</a>, quien lo acogió en la <a href="/wiki/Galia" title="Galia">Galia</a> como consejero legal y jefe de la oficina de comunicaciones. Apoyó a César en la <a href="/wiki/Segunda_guerra_civil_de_la_Rep%C3%BAblica_romana" title="Segunda guerra civil de la República romana">guerra civil</a> y tras su asesinato pasó al bando de Octavio (futuro <a href="/wiki/Augusto" title="Augusto">Augusto</a>), quien tuvo en gran estima a Trebacio hasta que este falleció en el 4&nbsp;d.&nbsp;C. Escribió varias obras relativas al <a href="/wiki/Derecho_romano" title="Derecho romano">Derecho romano</a>, pero ninguna se ha conservado.</p>` (with es->en model)
- Ignore certain tags, like `<code>` and `<samp>` #313 
- Never treat `<wbr>` as word-breaking #339 
- Allow for `<P></p>` (I haven't seen this go wrong but in theory it could have gone wrong.)
- That code that makes sure that each tag is returned in the output at least once? Yeah that was a bit too trigger happy. It now more correctly guarantees once _and only once_.
- Use the `isContinuation` logic to detect when to insert a space after an open or close tag. This will add a space in `un<u>der</u>line`, but not in `end of the <b>sentence</b>.` This change makes especially Wikipedia look more natural.
- Ignore tags like `<noscript>` at the parser level because [Firefox does so as well](https://searchfox.org/mozilla-central/source/parser/html/nsHtml5Tokenizer.cpp#176). Since Firefox can't guarantee that `<noscript>` tags have valid HTML inside them, this looks like the safest bet for now.

### CODE REVIEW

**Comment by jelmervdl** [comment]
I think this pull request has grown enough to start trying to clean it up and get it merged.

Preferably, #353 can be landed, before this gets merged. I will add some measures that show that this set of changes improves the HTML output quality. The description above mostly cover it: edge cases and refactoring to keep growing code complexity manageable.

I've tried to make HTML.cpp as readable as possible by leaving comments with my intentions throughout, but clang-tidy still gives be very, very low grades for certain functions (especially those with switch statements.) Please let me know if any bit of code's intended use isn't clear, or if you have suggestions on how I can split up certain parts to be more comprehensible.

**Comment by jerinphilip** [comment]
> Preferably, https://github.com/browsermt/bergamot-translator/pull/353 can be landed, before this gets merged. I will add some measures that show that this set of changes improves the HTML output quality.

Let's take this PR forward before #353. #353 looks like it will take longer - it's still exploratory. Let's discuss internally how to craft and position the numbers.

**Comment by jelmervdl** [comment]
I'll try to add a more elaborate test to bergamot-translator-tests to test some of the added edge cases as well.

File: src/translator/html.cpp [hunk 1]
```diff
@@ -622,7 +761,14 @@ void HTML::hardAlignments(Response const &response, std::vector<std::vector<size
...
 
-        if (currScore >= prevScore) {
+        TagStack const &currTagStack = sourceTokenSpans[offset + 1 + currSentenceIdx]->tags;
+        TagStack const &prevTagStack = sourceTokenSpans[offset + 1 + prevSentenceIdx]->tags;
+
+        // If this token has more markup, or a better score than the previous
+        // token (and they together are part of a word-ish thing) then mark
+        // this word as aligning. Otherwise just copy the alignment source of
+        // the previous token.
+        if (isSubset(prevTagStack, currTagStack) || currScore >= prevScore) {
```
Comment (jerinphilip) [inline src/translator/html.cpp:771]: Would appreciate some comments on the complexity of the overall process and relevant areas. This comparison is O(length), the tag comparisons can go full document length?

**Review by jerinphilip — COMMENTED** [review]
Leaving a few comments that are local to the source around in a first pass. I expect to take one more pass once my understanding of the HTML pipeline improves from exchanges here. 

I will try to do some black-box testing via the extension (as opposed to the previous python, thanks @jelmervdl for the dogfooding capabilities and improvements consequent of experimenting with the extension to the library).

Overall positive about this PR, so hoping to expedite merge. Most of the queries and suggestions below may be undertaken follow-up PR (save a suspected infinite loop).

File: src/translator/html.cpp [hunk 2]
```diff
@@ -68,30 +81,29 @@ std::string format(std::string const &formatTemplate, Arg arg, Args... args) {
...
-void diffTags(HTML::Taint const &prev, HTML::Taint const &curr, HTML::Taint &opening, HTML::Taint &closing) {
+/// When comparing two tag stacks, determine which tags need to be closed and
+/// opened to get from one stack to the other.
+void diffTags(HTML::TagStack const &prev, HTML::TagStack const &curr, HTML::TagStack &opening,
+              HTML::TagStack &closing) {
   opening.clear();
   closing.clear();
 
-  size_t i = 0;
+  std::int64_t i = 0;
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp:106]: According to clang: _Narrowing conversion from `size_t` (aka `unsigned long`) to signed type `std::__wrap_iter<marian::bergamot::HTML::Tag *const *>::difference_type` (aka `long`) is implementation-defined_

File: src/translator/html.cpp [hunk 3]
```diff
@@ -240,11 +282,85 @@ size_t debugCountTokens(AnnotatedText const &text) {
...
+      case markup::Scanner::TT_EOF:
+        ABORT("Did not find closing tag </{}>");
+      case markup::Scanner::TT_TAG_START:
+      case markup::Scanner::TT_TAG_END:
+        // Note: Looking specifically for only our own type of tag so we don't
+        // have to care about whether other tags we encounter are void tags or
+        // not. Does assume the HTML is valid, as no stack is kept.
+        if (toLowerCase(scanner.tag()) == name) {
+          if (token == markup::Scanner::TT_TAG_END) {
+            if (--inside == 0) break;  // also stops loop because !inside
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp:335]: `inside > 0` has to be true for the while-loop to start, so decrement should be safe. If decrement happens, it immediately jumps back to the `while (inside)` condition. I don't see a way for `--inside` to be called at that point.

File: src/translator/html.cpp [hunk 4]
```diff
@@ -108,10 +120,37 @@ bool intersects(ByteRange const &range, HTML::Span const &span) {
...
+
+bool contains(HTML::TagStack const &stack, HTML::Tag const *tag) {
   return std::find(stack.rbegin(), stack.rend(), tag) != stack.rend();
 }
 
+bool isSubset(HTML::TagStack const &a, HTML::TagStack const &b) {
+  if (a.size() > b.size()) return false;
+
+  for (auto i = a.begin(), j = b.begin(); i != a.end(); ++i, ++j)
+    if (*i != *j) return false;
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp:133]: It's a bad name indeed. It's more like isPrefix, i.e. is the markup in B an extension of the markup in A.

File: src/translator/html.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerinphilip) [reply src/translator/html.cpp:335]: May I request a simplify for the cognitive load (This is probably my dislike for evil pointer arithmetic in conditions showing, and preference towards compiler optimizing the statements if there's room).

```cpp
if (token == markup::Scanner::TT_TAG_END) {
    --inside; // Action which will happen when we enter this block, not as part of the condition below.
    if (inside == 0) { // condition only evaluates what it should evaluate
    	break;  // also stops loop because !inside
    }
    ...
}
```

[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp:335]: I'll change it.

Note that `inside` is just a counter, no pointer magic. I wrote it as `if inside counts down to 0, stop` but I admit that it only reads that way once you know to read it that way.

File: src/translator/xh_scanner.cpp [hunk 5]
```diff
@@ -312,11 +317,13 @@ bool Scanner::isWhitespace(char c) {
 
 Scanner::TokenType Scanner::scanComment() {
   if (gotTail_) {
+    start_ = input_.pos() - 3;  // minus "-->"
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/xh_scanner.cpp]: I'm using `length("whatever")` now in every place there is a magic constant, thanks!

File: src/translator/html.cpp [hunk 6]
```diff
@@ -108,10 +120,37 @@ bool intersects(ByteRange const &range, HTML::Span const &span) {
...
+size_t argmax(std::vector<T> const &items) {
+  assert(!items.empty());
+  size_t best = 0;
+  for (size_t i = 1; i + 1 < items.size(); ++i) {
+    if (items[i] > items[best]) {
+      best = i;
+    }
+  }
+  return best;
+}
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp]: I searched for argmax, didn't find it, assumed it wasn't in the standard library. Using max_element now (although the call site hasn't gotten any prettier because of it :( )

File: src/translator/html.cpp [hunk 7]
```diff
@@ -144,6 +182,7 @@ AnnotatedText apply(AnnotatedText const &in, Fun fun) {
   return out;
 }
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp]: Moved it to annotation.h (since it is a template)

File: src/translator/html.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jerinphilip) [reply src/translator/html.cpp:106]: Is it this line? I think we're treating the wrong symptom. Right below this, you're comparing this `i` to `.size()` for termination, so my understanding is this is to represent index into an addressable something. Hence `size_t`. 

Somewhere this appears to be compared to `std::__wrap_iter<marian::bergamot::HTML::Tag *const *>::difference_type` (range being `[-std::numeric_limits<size_t>::max(), -std::numeric_limits<size_t>::max()]`? Perhaps we should find that point and fix (I tried looking but not able to find immediately)?

[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp:106]: It's the lines below it where I add `i` to `prev.begin()` and `curr.begin()`. At least my cpp standard library expects a `difference_type` there for `random_acess_iterator::operator+(...)` so unsigned -> signed conversion.

Its all pretty pointless, `i` cannot be negative here, and I'm only ever adding. So yea I'm only listening to clang for advice here but personally `size_t` makes more sense to me.

[the remark this answers is not in the document]
Comment (jerinphilip) [reply src/translator/html.cpp:106]: Perhaps explicit cast at `prev.begin() + static_cast<long>(...)` to get rid of the warning, or perhaps don't bother as much and treat it as a guideline. I agree that overoptimizing here doesn't contribute much, and am on the same page regarding `size_t` making more sense (I hope we're on the same page `int64_t` is weird).

File: src/translator/html.h [hunk 8]
```diff
@@ -59,17 +70,17 @@ class HTML {
     // `attributes` and `data` with string_views pointing to it.
   };
 
-  using Taint = std::vector<Tag *>;
+  using TagStack = std::vector<Tag *>;
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.h:73]: I do like the intent reference_wrapper communicates, but I'm using pointer comparisons[1] and the nullability of bare pointers[2] in the code, which are not easily doable with reference_wrapper as far as I can see.

[1] [Do these two pointers paoint to the same tag, the memory location making a tag unique](https://github.com/browsermt/bergamot-translator/blob/a1ee8e9de3ecb757ac67ba1e7c715d0c7f9117f6/src/translator/html.cpp#L110)
[2]: [In HTML::HTML I have a pointer to the tag that's currently open, and which is null when it closes](https://github.com/browsermt/bergamot-translator/blob/a1ee8e9de3ecb757ac67ba1e7c715d0c7f9117f6/src/translator/html.cpp#L536)

File: src/translator/html.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerinphilip) [reply src/translator/html.cpp]: [`54be426` (#312)](https://github.com/browsermt/bergamot-translator/pull/312/commits/54be426b06943b6c05c283f977e05e2ca8c0f352) resolves this, looks like.

File: src/translator/html.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/html.cpp:137]: Yep! I renamed it to `extends` so now the code should read `stack B extends stack A`. That sounded like more familiar semantics to me.

File: src/translator/html.cpp (hunk 1 above)
Comment (jelmervdl) [reply src/translator/html.cpp]: Before we start: my complexity theory knowledge is … lacking. Please correct me if I'm wrong.

`isSubset` (now `extends`) is O(N), where N is the stack depth, but that stack pretty small on average. Maybe we should have a limit on how deep you can nest tags in the parser, just to avoid denial of service attacks.

This whole section I think is worst case O(N*N), where N in number of subwords in the document. Worst case is all subwords form a single word together and alignment scores keep getting higher as you go further down the document. On average again, the are only limited subwords per word so O(N).

The expensive bit is a bit higher where it looks for straggers, in `HTML::restoreTarget()`. That bit is O(N*N) where N is the number of spans (i.e. bits between tags) and that bit can use optimisations. A simple hashset would be sufficient there to get it back to O(N).

But putting this all in context, I've profiled various pages, both text heavy and tag heavy, and I've never seen the HTML parser even pop up in the top samples of the profiler. It's always either the model loading code (specifically `marian::binary::loadItems`) or the graph code that does the actual computations.

**Review by jerinphilip — APPROVED** [review]
Thanks for the extra tests. Let's get this in, and pursue things below in later PRs.

As feedback, that needn't necessarily be covered in this PR - someone trying to understand the HTML part will appreciate a bigger blob of documentation near the HTML class or something describing the components (parsing, what form of AST/DOM equivalent, target HTML gen/restoration) - at least I would.

Minor: I do not fully understand what *stragger* means (tried [googling](https://www.google.com/search?q=define%3Astragger)) in this context. I see most of the `Taint` (which I had troubled comprehending before) has been replaced with `TagStack`. 

Long term I think it will be a good idea to abstract different readers writers inserting something [pandoc ](https://pandoc.org/) like in between, thus providing a framework to translate formats (docx, markdown, LaTeX etc). Right now we're specializing for HTML.
